core/src/main/scala/kafka/utils/ZkUtils.scala

```diff
@@ -256,19 +251,41 @@ class ZkUtils(val zkClient: ZkClient,
   }
 
   /**
-   * Register brokers with v2 json format (which includes multiple endpoints).
+   * Register brokers with v3 json format (which includes multiple endpoints and rack).
```

Comment:
I think the reason that we have to design the broker registration json in a backward compatible way is that the old consumer and admin tools still read the json directly. Once we deprecate the old consumer and move the admin tools to the admin api, only the brokers need to read the json. Then, potentially we can make non compatible json changes as long as the broker writes the json according to inter.broker.protocol. For example, we make want to clean up the redundant host field in the future. So, it is true that at this moment, we can always write the latest version of the json that the broker understands since the change is backward compatible. However, following the inter.protocol convention is probably what we want to do in the future. So, we may want to just start doing this now.